README
NAME
    WebService::Linode - Perl Interface to the Linode.com API.

SYNOPSIS
        my $api = WebService::Linode->new( apikey => 'your api key here');
        print Dumper($api->linode_list);
        $api->linode_reboot(linodeid=>242);

    This module implements the Linode.com api methods. Linode methods have
    had dots replaced with underscores to generate the perl method name. All
    keys and parameters have been lower cased but returned data remains
    otherwise the same. For additional information see
    <http://www.linode.com/api/>

Constructor
    For documentation of possible arguments to the constructor, see
    WebService::Linode::Base.

Batch requests
    Each of the Linode API methods below may optionally be prefixed with
    QUEUE_ to add that request to a queue to be processed later in one or
    more batch requests which can be processed by calling
    send_queued_requests. For example:

        my @linode_ids = () # Get your linode ids through normal methods
        my @responses = map { $api->linode_ip_list( linodeid=>$_ ) } @linode_ids;

    Can be reduced to a single request:

        my @linode_ids = () # Get your linode ids through normal methods
        $api->QUEUE_linode_ip_list( linodeid=>$_ ) for @linode_ids;
        my @responses = $api->send_queued_requests; # One api request

    See WebService::Linode::Base for additional queue management methods.

   send_queued_requests
    Send queued batch requests, returns list of responses.

Methods from the Linode API
   avail_datacenters
   avail_nodebalancers
   avail_distributions
    Optional Parameters:

    *   distributionid

   avail_kernels
    Optional Parameters:

    *   isxen

    *   kernelid

   avail_linodeplans
    Optional Parameters:

    *   planid

   avail_stackscripts
    Optional Parameters:

    *   distributionid

    *   distributionvendor

    *   keywords

   domain_list
    Optional Parameters:

    *   domainid

   domain_update
    Required Parameters:

    *   domainid

    Optional Parameters:

    *   axfr_ips

    *   description

    *   domain

    *   expire_sec

    *   lpm_displaygroup

    *   master_ips

    *   refresh_sec

    *   retry_sec

    *   soa_email

    *   status

    *   ttl_sec

    *   type

   domain_create
    Required Parameters:

    *   domain

    *   type

    Optional Parameters:

    *   axfr_ips

    *   description

    *   expire_sec

    *   lpm_displaygroup

    *   master_ips

    *   refresh_sec

    *   retry_sec

    *   soa_email

    *   status

    *   ttl_sec

   domain_delete
    Required Parameters:

    *   domainid

   domain_resource_update
    Required Parameters:

    *   resourceid

    Optional Parameters:

    *   domainid

    *   name

    *   port

    *   priority

    *   protocol

    *   target

    *   ttl_sec

    *   weight

   domain_resource_list
    Required Parameters:

    *   domainid

    Optional Parameters:

    *   resourceid

   domain_resource_delete
    Required Parameters:

    *   domainid

    *   resourceid

   domain_resource_create
    Required Parameters:

    *   domainid

    *   type

    Optional Parameters:

    *   name

    *   port

    *   priority

    *   protocol

    *   target

    *   ttl_sec

    *   weight

   linode_resize
    Required Parameters:

    *   linodeid

    *   planid

   linode_list
    Optional Parameters:

    *   linodeid

   linode_mutate
    Required Parameters:

    *   linodeid

   linode_boot
    Required Parameters:

    *   linodeid

    Optional Parameters:

    *   configid

   linode_create
    Required Parameters:

    *   datacenterid

    *   planid

    Optional Parameters:

    *   paymentterm

   linode_clone
    Required Parameters:

    *   datacenterid

    *   linodeid

    *   planid

    Optional Parameters:

    *   paymentterm

   linode_update
    Required Parameters:

    *   linodeid

    Optional Parameters:

    *   alert_bwin_enabled

    *   alert_bwin_threshold

    *   alert_bwout_enabled

    *   alert_bwout_threshold

    *   alert_bwquota_enabled

    *   alert_bwquota_threshold

    *   alert_cpu_enabled

    *   alert_cpu_threshold

    *   alert_diskio_enabled

    *   alert_diskio_threshold

    *   backupweeklyday

    *   backupwindow

    *   label

    *   lpm_displaygroup

    *   ms_ssh_disabled

    *   ms_ssh_ip

    *   ms_ssh_port

    *   ms_ssh_user

    *   watchdog

   linode_webconsoletoken
    Required Parameters:

    *   linodeid

   linode_reboot
    Required Parameters:

    *   linodeid

    Optional Parameters:

    *   configid

   linode_shutdown
    Required Parameters:

    *   linodeid

   linode_delete
    Required Parameters:

    *   linodeid

    Optional Parameters:

    *   skipchecks

   linode_config_delete
    Required Parameters:

    *   configid

    *   linodeid

   linode_config_create
    Required Parameters:

    *   kernelid

    *   label

    *   linodeid

    Optional Parameters:

    *   comments

    *   devtmpfs_automount

    *   disklist

    *   helper_depmod

    *   helper_disableupdatedb

    *   helper_network

    *   helper_xen

    *   ramlimit

    *   rootdevicecustom

    *   rootdevicenum

    *   rootdevicero

    *   runlevel

   linode_config_update
    Required Parameters:

    *   configid

    Optional Parameters:

    *   comments

    *   devtmpfs_automount

    *   disklist

    *   helper_depmod

    *   helper_disableupdatedb

    *   helper_network

    *   helper_xen

    *   kernelid

    *   label

    *   linodeid

    *   ramlimit

    *   rootdevicecustom

    *   rootdevicenum

    *   rootdevicero

    *   runlevel

   linode_config_list
    Required Parameters:

    *   linodeid

    Optional Parameters:

    *   configid

   linode_disk_createfromimage
    Required Parameters:

    *   imageid

    *   linodeid

    Optional Parameters:

    *   label

    *   rootpass

    *   rootsshkey

    *   size

   linode_disk_duplicate
    Required Parameters:

    *   diskid

    *   linodeid

   linode_disk_update
    Required Parameters:

    *   diskid

    Optional Parameters:

    *   isreadonly

    *   label

    *   linodeid

   linode_disk_createfromstackscript
    Required Parameters:

    *   distributionid

    *   label

    *   linodeid

    *   rootpass

    *   size

    *   stackscriptid

    *   stackscriptudfresponses

    Optional Parameters:

    *   rootsshkey

   linode_disk_imagize
    Required Parameters:

    *   diskid

    *   linodeid

    Optional Parameters:

    *   description

    *   label

   linode_disk_delete
    Required Parameters:

    *   diskid

    *   linodeid

   linode_disk_resize
    Required Parameters:

    *   diskid

    *   linodeid

    *   size

   linode_disk_list
    Required Parameters:

    *   linodeid

    Optional Parameters:

    *   diskid

   linode_disk_createfromdistribution
    Required Parameters:

    *   distributionid

    *   label

    *   linodeid

    *   rootpass

    *   size

    Optional Parameters:

    *   rootsshkey

   linode_disk_create
    Required Parameters:

    *   label

    *   linodeid

    *   size

    *   type

    Optional Parameters:

    *   fromdistributionid

    *   isreadonly

    *   rootpass

    *   rootsshkey

   linode_ip_setrdns
    Required Parameters:

    *   hostname

    *   ipaddressid

   linode_ip_swap
    Required Parameters:

    *   ipaddressid

    Optional Parameters:

    *   tolinodeid

    *   withipaddressid

   linode_ip_addprivate
    Required Parameters:

    *   linodeid

   linode_ip_list
    Required Parameters:

    *   linodeid

    Optional Parameters:

    *   ipaddressid

   linode_ip_addpublic
    Required Parameters:

    *   linodeid

   linode_job_list
    Required Parameters:

    *   linodeid

    Optional Parameters:

    *   jobid

    *   pendingonly

   image_delete
    Required Parameters:

    *   imageid

   image_list
    Optional Parameters:

    *   imageid

    *   pending

   image_update
    Required Parameters:

    *   imageid

    Optional Parameters:

    *   label

    *   description

   stackscript_list
    Optional Parameters:

    *   stackscriptid

   stackscript_update
    Required Parameters:

    *   stackscriptid

    Optional Parameters:

    *   description

    *   distributionidlist

    *   ispublic

    *   label

    *   rev_note

    *   script

   stackscript_create
    Required Parameters:

    *   distributionidlist

    *   label

    *   script

    Optional Parameters:

    *   description

    *   ispublic

    *   rev_note

   stackscript_delete
    Required Parameters:

    *   stackscriptid

   nodebalancer_config_delete
    Required Parameters:

    *   configid

    *   nodebalancerid

   nodebalancer_config_create
    Required Parameters:

    *   nodebalancerid

    Optional Parameters:

    *   algorithm

    *   check

    *   check_attempts

    *   check_body

    *   check_interval

    *   check_path

    *   check_timeout

    *   port

    *   protocol

    *   ssl_cert

    *   ssl_key

    *   stickiness

   nodebalancer_config_update
    Required Parameters:

    *   configid

    Optional Parameters:

    *   algorithm

    *   check

    *   check_attempts

    *   check_body

    *   check_interval

    *   check_path

    *   check_timeout

    *   port

    *   protocol

    *   ssl_cert

    *   ssl_key

    *   stickiness

   nodebalancer_config_list
    Required Parameters:

    *   nodebalancerid

    Optional Parameters:

    *   configid

   nodebalancer_node_create
    Required Parameters:

    *   address

    *   configid

    *   label

    Optional Parameters:

    *   mode

    *   weight

   nodebalancer_node_delete
    Required Parameters:

    *   nodeid

   nodebalancer_node_list
    Required Parameters:

    *   configid

    Optional Parameters:

    *   nodeid

   nodebalancer_node_update
    Required Parameters:

    *   nodeid

    Optional Parameters:

    *   address

    *   label

    *   mode

    *   weight

   user_getapikey
    Required Parameters:

    *   password

    *   username

    Optional Parameters:

    *   expires

    *   label

    *   token

AUTHORS
    *   Michael Greb, "<michael@thegrebs.com>"

    *   Stan "The Man" Schwertly "<stan@schwertly.com>"

COPYRIGHT & LICENSE
    Copyright 2008-2014 Michael Greb, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl module WebService::Linode

0.26  2015-01-06 20:01:46 EST
    - Patch from James Ottinger for image_update method 74816d

0.25  2014-12-10 12:52:40 PST
    - Patch from James Ottinger for additional Linode Images support a89bfc
    - support for network helper in linode_config_create & linode_config_update
      see <https://forum.linode.com/viewtopic.php?f=26&t=11469>

0.24  2014-10-18 22:43:33 EDT
    - add explicit section to POD for send_queued_requests
    - Add Test::Pod::Coverage to cpanfile for develop

0.23  2014-10-14 16:31:36 EDT
    - Batched API request support by Kevan Benson <kbenson@cpan.org>

0.22  2014-07-24 10:13:47 EDT
    - support for the Linode Images service, currently in beta
      see <https://forum.linode.com/viewtopic.php?f=26&t=11180>

0.21  2014-07-01 13:02:10 EDT
    - new avail_nodebalancers and linode_ip_setrdns methods
    - linode_disk_create now optionaly takes fromdistributionid, isreadonly,
      rootpass, and rootsshkey
    - linode_disk_createfromstackscript now optionally takes rootsshkey
    - paymentterm is now optional for nodebalancer_create

0.20  2014-03-13 11:42:49 EDT
    - support for adding new private and public IPs to a Linode
    - support for swapping/moving an IP between Linodes
    - moar deprecated WebService::Linode::DNS
      - all methods carp
      - most of POD replaced, points to WebService::Linode
    - add LWP to cpanfile for CPANTS

0.19  2014-02-09 00:28:23 EST
    - depend on just LWP::Protocol::https instead of LWP::UserAgent,
      Crypt::SSLeay, & Mozilla::CA, thanks Christopher Meng (cicku)
    - Update FSF mailing address in LICENSE, thanks Christopher Meng (cicku)

0.18  2014-02-04 11:36:24 EST
    - API 2 factor authentication support, see:
    <https://blog.linode.com/2014/02/03/linode-api-multiple-keys-and-two-factor/>

0.17  2014-01-24 09:30:48 EST
    - remove deprecated getDomainIDbyName & getDomainResourceIDbyName
    - allow setting API url via ENV var LINODE_API_URL
    - paymentterm is optional for linode.create and linode.clone
    - update examples/dyndns.pl to remove use of deprecated functions

0.16  2014-01-06 09:59:02 EST
    - useragent arg to constructor for custom user agent header

0.15  2013-11-07 12:02:21 EST
    - Treat error code 0 as OK
    - Update to latest API spec:
        - New optional argument for domain_create & domain_update:
            lpm_displaygroup
        - New optional arguments for nodebalancer_config_create and
          nodebalancer_config_update:
            ssl_cert, ssl_key
    - Add deprecation warning when calling getDomainIDbyName &
      getDomainResourceIDbyName

0.14  2013-10-23 11:04:39 EDT
    - Add Mozilla::CA to dependencies.

0.13  2013-09-28 20:05:23 EDT
    - No functional changes in this release.
    - Update link to Linode API docs
    - Clean up POD a bit, remove required/optional parameter headings when list
      is empty
    - https://github.com/mikegrb/WebService-Linode/pull/2 to update older
      change log entries to CPAN::Changes::Spec, thanks @neilbowers

0.12  2013-09-25 14:58:18 EDT
    - Upstream API Changes
        - New methods:
            - Prefixed by account_: estimateinvoice, paybalance, updatecard
            - Prefixed by linode_:  mutate, webconsoletoken
        - New optional args for linode_update:
            - ms_ssh_disabled, ms_ssh_ip, ms_ssh_port, ms_ssh_user
    - Bugfix
        - fix avail_plans requiring optional planid argument
        - verify domain_resource_create and linode_config_create still require
          primary key as avail_plans does in api.spec
    - Dist Creation Changes
        - converted dist creation to milla
        - moved helper scripts to
          https://github.com/mikegrb/WebService-Linode-AuthorUtils

0.11 2012-02-11
    - fix primary keys marked as required param for domain_resource_create and
      linode_config_create (thanks Peter Stephan)
      - note: not fixed upstream yet gen-validation.pl output can not be used
        directly currently

0.10 2012-20-08
    - updated API definition from api.spec after Linode fix, closes #79093

0.09 2012-29-07
    - Forgot to bump version number in DNS.pm and Base.pm
        this is why all my other modules use Dist::Zilla

0.08 2012-29-07
    - new upstream API calls added
        many additions including nodebalancer support
        see for full changes http://j.mp/MNKKTf
    - patch from stephon for deprecated DNS.pm (closes rt #71833)

0.07 2012-10-02
    - Correct typo in resourceid argument to domain_resource->list

0.06 1012-12-01
    - Correct reference to module name in docs (closes rt #59535 reported
      by Paul Hansen)
    - Use direct object notation rather than indirect notation
    - new upstream API calls added

0.05 2009-09-14
    - Apikey is no longer a required argument to the base constructor
    - New method in the base class, apikey for specifying or changing apikey
      at times other than construction
    - New method in WebService::Linode, user_getapikey, given a valid Linode
      username and password, returns $data->{api_key}.  Note: you are
      currently still responsible for calling $api->apikey($key) after
      retrieving the key if you wish to use it for subsequent requests.
    - fix WebService::Linode::DNS domainResourceGet when called with domain
      name rather than id. (closes #49452 reported by jtmink1)

0.04 2009-08-10 21:45:49 EDT
    - Update to Linode.com API version 2.0 -- This is a pretty big change,
      includes support for Linodes now as well as DNS
    - WebService::Linode::DNS is now deprecated but /should/ continue to work.
      changing to WebService::Linode should be straight forward with a few
      method names changed to match the Linode.com methods with dots replaced
      with under scores.

0.03 2008-12-10
    - Patch from Pat Hennessy to use $response->status_line for error
      message when no JSON found in the response (closes #41328)

lib/WebService/Linode.pm
package WebService::Linode;

require 5.006000;

use warnings;
use strict;

use Carp;
use List::Util qw(first);
use WebService::Linode::Base;

our $VERSION = '0.07';
our @ISA     = ("WebService::Linode::Base");
our $AUTOLOAD;

my %validation = (
    avail => {
        datacenters   => [ [], [                     ] ],
        kernels       => [ [], [ 'kernelid', 'isxen' ] ],
        linodeplans   => [ [], [ 'plainid'           ] ],
        distributions => [ [], [ 'distributionid'    ] ],
        stackscripts  => [ [], [ 'distributionid', 'distributionvendor', 'keywords'] ],
    },
    domain => {
        create => [ [ 'domain', 'type' ], [ qw( description soa_email refresh_sec retry_sec expire_sec ttl_sec status master_ips ) ]],
        list   => [ [], ['domainid'] ],
        update => [ ['domainid'], [ qw( domain description type soa_email refresh_sec retry_sec expire_sec ttl_sec  status master_ips) ]],
        delete => [ ['domainid'], [] ],
    },
    domain_resource => {
        create => [ ['domainid', 'type'], [ qw( name target priority weight port protocol ttl_sec )] ],
        list   => [ ['domainid'], ['resourceid'] ],
        update => [ [ 'domainid', 'resourceid' ], [ qw( name target priority weight port protocol ttl_sec )] ],
        delete => [ [ 'domainid', 'resourceid' ], [] ],
    },
    linode => {
        create   => [ [ qw( datacenterid planid paymentterm ) ], [] ],
        list     => [ [], ['linodeid'] ],
        update   => [ ['linodeid'], [ qw( label lpm_displaygroup alert_cpu_enabled alert_cpu_threshold alert_diskio_enabled alert_diskio_threshold alert_bwin_enabled alert_bwin_threshold alert_bwout_enabled alert_bwout_threshold alert_bwquota_enabled alert_bwquota_threshold backupwindow backupweeklyday watchdog ) ] ],
        delete   => [ ['linodeid'], [] ],
        shutdown => [ ['linodeid'], [] ],
        boot     => [ ['linodeid'], ['configid'] ],
        reboot   => [ ['linodeid'], ['configid'] ],
        resize   => [ ['linodeid', 'planid'], [] ],
    },

    linode_config => {
        create => [ [ qw( linodeid kernelid label ) ], [ qw( comments ramlimit disklist runlevel rootdevicenum rootdevicecustom rootdevicero helper_disableupdatedb helper_xen helper_depmod ) ] ],
        list   => [ ['linodeid'], ['configid'] ],
        update => [ [ 'linodeid', 'configid' ],  [ qw( kernelid label comments ramlimit disklist runlevel rootdevicenum rootdevicecustom rootdevicero helper_xen helper_disableupdatedb helper_depmod ) ] ],
        delete => [ [ 'linodeid', 'configid' ], [] ],

    },

    linode_disk => {
        create    => [ [ 'linodeid', 'label', 'type', 'size' ], [ 'isreadonly' ] ],
        list      => [ [ 'linodeid' ], ['diskid' ] ],
        update    => [ [ 'linodeid', 'diskid' ], [ 'label', 'isreadonly' ] ],
        delete    => [ [ 'linodeid', 'diskid' ], [] ],
        createfromdistribution => [ [ qw( linodeid distributionid label size rootpass ) ], [ 'rootsshkey' ] ],
        duplicate => [ [ 'linodeid', 'diskid' ], [] ],
        resize    => [ [ 'linodeid', 'diskid', 'size' ], [] ],
        createfromstackscript  => [ [ qw( linodeid stackscriptid stackscriptudfresponses distributionid label size rootpass) ], [] ],

    },
    linode_ip => {
        list       => [ [ 'linodeid' ], [ 'ipaddressid' ] ],
        addprivate => [ [ 'linodeid' ], [] ],
    },
    linode_job => {
        list => [ [ 'linodeid' ], [ 'jobid', 'pendingonly' ] ],
    },
    stackscript => {
	    create => [ ['label', 'distributionidlist', 'script' ], ['description', 'ispublic', 'rev_note'] ],
        delete => [ ['stackscriptid'], [] ],
        list   => [ ['stackscriptid'], [] ],
        update => [ ['stackscriptid'], [ qw( label description distributionidlist ispublic rev_note script) ] ],
    },
    test => {
	    echo   => [ [], [] ],
    },
    user => {
        getapikey => [ [ 'username', 'password' ], [] ],
    },
    account => {
        info => [ [], [] ],
    },
);

sub AUTOLOAD {
    (my $name = $AUTOLOAD) =~ s/.+:://;
    return if $name eq 'DESTROY';
    if ($name =~ m/^(.*?)_([^_]+)$/) {
        my ($thing, $action) = ($1, $2);
        if (exists $validation{$thing} && exists $validation{$thing}{$action}) {
            no strict 'refs';
            *{ $AUTOLOAD } = sub {
                my ($self, %args) = @_;
                for my $req ( @{ $validation{$thing}{$action}[0] } ) {
                    if ( !exists $args{$req} ) {
                        carp "Missing required argument $req for ${thing}_${action}";
                        return;
                    }
                }
                for my $given ( keys %args ) {
                    if (!first { $_ eq $given }
                        @{ $validation{$thing}{$action}[0] },
                        @{ $validation{$thing}{$action}[1] } )
                    {   carp "Unknown argument $given for ${thing}_${action}";
                        return;
                    }
                }
                (my $apiAction = "${thing}_${action}") =~ s/_/./g;
                my $data = $self->do_request( api_action => $apiAction, %args);
                return [ map { $self->_lc_keys($_) } @$data ] if ref $data eq 'ARRAY';
                return $self->_lc_keys($data) if ref $data eq 'HASH';
                return $data;
            };
            goto &{ $AUTOLOAD };
        }
        else {
            carp "Can't call ${thing}_${action}";
            return;
        }
        return;
    }
    croak "Undefined subroutine \&$AUTOLOAD called";
}

sub getDomainIDbyName {
    my ($self, $name) = @_;
    foreach my $domain (@{$self->domain_list()}) {
        return $domain->{domainid} if $domain->{domain} eq $name;
    }
    return;
}

sub getDomainResourceIDbyName {
    my ( $self, %args ) = @_;
    $self->_debug( 10, 'getResourceIDbyName called' );

    my $domainid = $args{domainid};
    if ( !exists( $args{domainid} ) && exists( $args{domain} ) ) {
        $domainid = $self->getDomainIDbyName( $args{domain} );
    }

    if ( !( defined($domainid) && exists( $args{name} ) ) ) {
        $self->_error( -1,
            'Must pass domain or domainid and (resource) name to getResourceIDbyName'
        );
        return;
    }

    for my $rr ( @{ $self->domain_resource_list( domainid => $domainid ) } ) {
        return $rr->{resourceid} if $rr->{name} eq $args{name};
    }
}

'mmm, cake';
__END__

=head1 NAME

WebService::Linode - Perl Interface to the Linode.com API.

=head1 VERSION

Version 0.07

=head1 SYNOPSIS

    my $api = WebService::Linode->new( apikey => 'your api key here');
    print Dumper($api->linode_list);
    $api->linode_reboot(linodeid=>242);

This module implements the Linode.com api methods.  Linode methods have had
dots replaced with underscores to generate the perl method name.  All keys
and parameters have been lower cased but returned data remains otherwise the
same.  For additional information see L<http://www.linode.com/api/autodoc.cfm>

=head1 Methods from the Linode API

=head3 avail_stackscripts

Required Parameters:

=over 4

=back

Optional Parameters:

=over 4

=item * distributionid

=item * distributionvendor

=item * keywords

=back

=head3 avail_kernels

Required Parameters:

=over 4

=back

Optional Parameters:

=over 4

=item * kernelid

=item * isxen

=back

=head3 avail_linodeplans

Required Parameters:

=over 4

=back

Optional Parameters:

=over 4

=item * plainid

=back

=head3 avail_datacenters

Required Parameters:

=over 4

=back

Optional Parameters:

=over 4

=back

=head3 avail_distributions

Required Parameters:

=over 4

=back

Optional Parameters:

=over 4

=item * distributionid

=back

=head3 domain_create

Required Parameters:

=over 4

=item * domain

=item * type

=back

Optional Parameters:

=over 4

=item * description

=item * soa_email

=item * refresh_sec

=item * retry_sec

=item * expire_sec

=item * ttl_sec

=item * status

=item * master_ips

=back

=head3 domain_delete

Required Parameters:

=over 4

=item * domainid

=back

Optional Parameters:

=over 4

=back

=head3 domain_update

Required Parameters:

=over 4

=item * domainid

=back

Optional Parameters:

=over 4

=item * domain

=item * description

=item * type

=item * soa_email

=item * refresh_sec

=item * retry_sec

=item * expire_sec

=item * ttl_sec

=item * status

=item * master_ips

=back

=head3 domain_list

Required Parameters:

=over 4

=back

Optional Parameters:

=over 4

=item * domainid

=back

=head3 domain_resource_create

Required Parameters:

=over 4

=item * domainid

=item * type

=back

Optional Parameters:

=over 4

=item * name

=item * target

=item * priority

=item * weight

=item * port

=item * protocol

=item * ttl_sec

=back

=head3 domain_resource_delete

Required Parameters:

=over 4

=item * domainid

=item * resourceid

=back

Optional Parameters:

=over 4

=back

=head3 domain_resource_update

Required Parameters:

=over 4

=item * domainid

=item * resourceid

=back

Optional Parameters:

=over 4

=item * name

=item * target

=item * priority

=item * weight

=item * port

=item * protocol

=item * ttl_sec

=back

=head3 domain_resource_list

Required Parameters:

=over 4

=item * domainid

=back

Optional Parameters:

=over 4

=item * resourceid

=back

=head3 linode_create

Required Parameters:

=over 4

=item * datacenterid

=item * planid

=item * paymentterm

=back

Optional Parameters:

=over 4

=back

=head3 linode_reboot

Required Parameters:

=over 4

=item * linodeid

=back

Optional Parameters:

=over 4

=item * configid

=back

=head3 linode_boot

Required Parameters:

=over 4

=item * linodeid

=back

Optional Parameters:

=over 4

=item * configid

=back

=head3 linode_resize

Required Parameters:

=over 4

=item * linodeid

=item * planid

=back

Optional Parameters:

=over 4

=back

=head3 linode_shutdown

Required Parameters:

=over 4

=item * linodeid

=back

Optional Parameters:

=over 4

=back

=head3 linode_delete

Required Parameters:

=over 4

=item * linodeid

=back

Optional Parameters:

=over 4

=back

=head3 linode_update

Required Parameters:

=over 4

=item * linodeid

=back

Optional Parameters:

=over 4

=item * label

=item * lpm_displaygroup

=item * alert_cpu_enabled

=item * alert_cpu_threshold

=item * alert_diskio_enabled

=item * alert_diskio_threshold

=item * alert_bwin_enabled

=item * alert_bwin_threshold

=item * alert_bwout_enabled

=item * alert_bwout_threshold

=item * alert_bwquota_enabled

=item * alert_bwquota_threshold

=item * backupwindow

=item * backupweeklyday

=item * watchdog

=back

=head3 linode_list

Required Parameters:

=over 4

=back

Optional Parameters:

=over 4

=item * linodeid

=back

=head3 linode_config_create

Required Parameters:

=over 4

=item * linodeid

=item * kernelid

=item * label

=back

Optional Parameters:

=over 4

=item * comments

=item * ramlimit

=item * disklist

=item * runlevel

=item * rootdevicenum

=item * rootdevicecustom

=item * rootdevicero

=item * helper_disableupdatedb

=item * helper_xen

=item * helper_depmod

=back

=head3 linode_config_delete

Required Parameters:

=over 4

=item * linodeid

=item * configid

=back

Optional Parameters:

=over 4

=back

=head3 linode_config_update

Required Parameters:

=over 4

=item * linodeid

=item * configid

=back

Optional Parameters:

=over 4

=item * kernelid

=item * label

=item * comments

=item * ramlimit

=item * disklist

=item * runlevel

=item * rootdevicenum

=item * rootdevicecustom

=item * rootdevicero

=item * helper_xen

=item * helper_disableupdatedb

=item * helper_depmod

=back

=head3 linode_config_list

Required Parameters:

=over 4

=item * linodeid

=back

Optional Parameters:

=over 4

=item * configid

=back

=head3 linode_disk_create

Required Parameters:

=over 4

=item * linodeid

=item * label

=item * type

=item * size

=back

Optional Parameters:

=over 4

=item * isreadonly

=back

=head3 linode_disk_createfromstackscript

Required Parameters:

=over 4

=item * linodeid

=item * stackscriptid

=item * stackscriptudfresponses

=item * distributionid

=item * label

=item * size

=item * rootpass

=back

Optional Parameters:

=over 4

=back

=head3 linode_disk_createfromdistribution

Required Parameters:

=over 4

=item * linodeid

=item * distributionid

=item * label

=item * size

=item * rootpass

=back

Optional Parameters:

=over 4

=item * rootsshkey

=back

=head3 linode_disk_resize

Required Parameters:

=over 4

=item * linodeid

=item * diskid

=item * size

=back

Optional Parameters:

=over 4

=back

=head3 linode_disk_duplicate

Required Parameters:

=over 4

=item * linodeid

=item * diskid

=back

Optional Parameters:

=over 4

=back

=head3 linode_disk_delete

Required Parameters:

=over 4

=item * linodeid

=item * diskid

=back

Optional Parameters:

=over 4

=back

=head3 linode_disk_update

Required Parameters:

=over 4

=item * linodeid

=item * diskid

=back

Optional Parameters:

=over 4

=item * label

=item * isreadonly

=back

=head3 linode_disk_list

Required Parameters:

=over 4

=item * linodeid

=back

Optional Parameters:

=over 4

=item * diskid

=back

=head3 linode_ip_addprivate

Required Parameters:

=over 4

=item * linodeid

=back

Optional Parameters:

=over 4

=back

=head3 linode_ip_list

Required Parameters:

=over 4

=item * linodeid

=back

Optional Parameters:

=over 4

=item * ipaddressid

=back

=head3 linode_job_list

Required Parameters:

=over 4

=item * linodeid

=back

Optional Parameters:

=over 4

=item * jobid

=item * pendingonly

=back

=head3 stackscript_create

Required Parameters:

=over 4

=item * label

=item * distributionidlist

=item * script

=back

Optional Parameters:

=over 4

=item * description

=item * ispublic

=item * rev_note

=back

=head3 stackscript_delete

Required Parameters:

=over 4

=item * stackscriptid

=back

Optional Parameters:

=over 4

=back

=head3 stackscript_update

Required Parameters:

=over 4

=item * stackscriptid

=back

Optional Parameters:

=over 4

=item * label

=item * description

=item * distributionidlist

=item * ispublic

=item * rev_note

=item * script

=back

=head3 stackscript_list

Required Parameters:

=over 4

=item * stackscriptid

=back

Optional Parameters:

=over 4

=back

=head1 Additional Helper Methods

=head3 getDomainIDbyName( domain => 'example.com' )

Returns the ID for a domain given the name.

=head3 getDomainResourceIDbyName( domainid => 242, name => 'www')

Takes a record name and domainid or domain and returns the resourceid.

=head1 AUTHORS

=over

=item * Michael Greb, C<< <mgreb@linode.com> >>

=item * Stan "The Man" Schwertly C<< <stan@linode.com> >>

=back

=head1 BUGS

This module does not yet support the Linode API batch method, patches welcome.

Please report any bugs or feature requests to C<bug-webservice-linode
at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebService-Linode>.  I will
be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WebService::Linode


You can also look for information at:

=over 4

=item * Module Repo

L<http://git.thegrebs.com/?p=WebService-Linode;a=summary>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WebService-Linode>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WebService-Linode>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WebService-Linode>

=item * Search CPAN

L<http://search.cpan.org/dist/WebService-Linode>

=back

=head1 COPYRIGHT & LICENSE

Copyright 2008-2009 Linode, LLC, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

